### ISSUE TITLE
CMake and cargo cannot detect Clang 13.0.1 from mingw-w64-clang-x86_64-clang because it lacks the necessary symlinks

### ISSUE THREAD

**Issue by amyspark** [msys2/MINGW-packages#10800]
Hi,

Some time between 13.0.0-3 and 13.0.1-1, the gcc and g++ "symlinks" were removed from the `mingw-w64-clang-x86_64-clang` package:

```
$ tar tvf mingw-w64-clang-x86_64-clang-12.0.1-5-any.pkg.tar.zst  | grep g++
-rwxr-xr-x root/root    154112 2021-08-20 09:08 clang64/bin/clang++.exe
-rwxr-xr-x root/root    154112 2021-08-20 09:08 clang64/bin/g++.exe
-rwxr-xr-x root/root    154112 2021-08-20 09:08 clang64/bin/x86_64-w64-mingw32-clang++.exe
-rwxr-xr-x root/root    154112 2021-08-20 09:08 clang64/bin/x86_64-w64-mingw32-g++.exe
```

```
$ tar tvf mingw-w64-clang-x86_64-clang-13.0.0-3-any.pkg.tar.zst  | grep g++
-rwxr-xr-x root/root    151040 2021-10-31 21:45 clang64/bin/clang++.exe
-rwxr-xr-x root/root    151040 2021-10-31 21:45 clang64/bin/g++.exe
-rwxr-xr-x root/root    151040 2021-10-31 21:45 clang64/bin/x86_64-w64-mingw32-clang++.exe
-rwxr-xr-x root/root    151040 2021-10-31 21:45 clang64/bin/x86_64-w64-mingw32-g++.exe
```
 
```
$ tar tvf mingw-w64-clang-x86_64-clang-13.0.1-1-any.pkg.tar.zst  | grep g++
-rwxr-xr-x root/root    151040 2022-02-04 15:50 clang64/bin/clang++.exe
-rwxr-xr-x root/root    151040 2022-02-04 15:50 clang64/bin/x86_64-w64-mingw32-clang++.exe

```

This completely breaks the ability to use the CLANG64 toolchain in my scripted builds, as CMake (currently 3.22.2) only detects compilers with  `gcc` or `g++` executable names.

This was caused by: https://github.com/msys2/MINGW-packages/commit/e13b81aaeb6b8401829ac335f15e5643c012a7d2, which removed the needed symlinks from the list.

https://github.com/msys2/MINGW-packages/blob/6964600c7cf85840d22403e9fd68135f39090534/mingw-w64-clang/PKGBUILD#L199

https://github.com/msys2/MINGW-packages/blob/e13b81aaeb6b8401829ac335f15e5643c012a7d2/mingw-w64-clang/PKGBUILD#L199

**Comment by jeremyd2019** [comment]
Discussion about this change: https://github.com/msys2/MINGW-packages/pull/10651#discussion_r798529212

**Comment by mati865** [comment]
I'm not against bringing back GCC named wrappers but doesn't CMake respect `CC` and `CXX` flags?

**Comment by MehdiChinoune** [comment]
> This completely breaks the ability to use the CLANG64 toolchain in my scripted builds, as CMake (currently 3.22.2) only detects compilers with `gcc` or `g++` executable names.

I don't think so, CMake search for other compilers when It doesn't find gcc/g++.
Could you show a reproducer!

**Comment by amyspark** [comment]
> -- Building for: Ninja

You're using the Ninja generator, which indeed works. Compare with `-G "MinGW Makefiles"` without the `CC`/`CXX` trick:

```
[proc] Executing command: "C:\Program Files\CMake\bin\cmake.EXE" --no-warn-unused-cli -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -DCMAKE_BUILD_TYPE:STRING=Debug -DCMAKE_INSTALL_PREFIX:STRING=c:/Users/Amalia/Desktop/test/install -Hc:/Users/Amalia/Desktop/test -Bc:/Users/Amalia/Desktop/test/build -G "MinGW Makefiles"
[cmake] Not searching for unused variables given on the command line.
[cmake] -- The C compiler identification is unknown
[cmake] -- The CXX compiler identification is unknown
[cmake] CMake Error at CMakeLists.txt:1 (project):
[cmake]   The CMAKE_C_COMPILER:
[cmake] 
[cmake]     gcc.exe
[cmake] 
[cmake]   is not a full path and was not found in the PATH.
[cmake] 
[cmake]   Tell CMake where to find the compiler by setting either the environment
[cmake]   variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
[cmake]   the compiler, or to the compiler name if it is in the PATH.
[cmake] 
[cmake] 
[cmake] CMake Error at CMakeLists.txt:1 (project):
[cmake]   The CMAKE_CXX_COMPILER:
[cmake] 
[cmake]     g++.exe
[cmake] 
[cmake]   is not a full path and was not found in the PATH.
[cmake] 
[cmake]   Tell CMake where to find the compiler by setting either the environment
[cmake]   variable "CXX" or the CMake cache entry CMAKE_CXX_COMPILER to the full path
[cmake]   to the compiler, or to the compiler name if it is in the PATH.
[cmake] 
[cmake] 
[cmake] -- Configuring incomplete, errors occurred!
[cmake] See also "C:/Users/Amalia/Desktop/test/build/CMakeFiles/CMakeOutput.log".
[cmake] See also "C:/Users/Amalia/Desktop/test/build/CMakeFiles/CMakeError.log".
```

This is my Visual Studio Code configuration:

```json
{
    "cmake.configureEnvironment": {
        "PATH": "D:/msys64/clang64/bin;${env:PATH}"
    },
    "cmake.sourceDirectory": "${workspaceFolder}/.",
    "cmake.generator": "MinGW Makefiles"
}
```

**Comment by Biswa96** [comment]
The `C:\Program Files\CMake\bin\cmake.EXE` program is not part of msys2/mingw environment. Try with the mingw CMake package from msys2.

**Comment by amyspark** [comment]
> Try with the mingw CMake package from msys2.

We are not authorized to use msys2 packages except for the bare toolchain. Again, this use case did work until the commit I referenced above.

Do notice too that this affects Boost builds with the `toolset=gcc` parameter; we now need to add another override to target Clang and its specific suffix to make our ExternalProject work.

**Comment by MehdiChinoune** [comment]
I can confirm with `MinGW Makefiles`:
```
$ cmake ../.. -G"MinGW Makefiles"
-- The CXX compiler identification is unknown
-- The C compiler identification is unknown
CMake Error at CMakeLists.txt:9 (project):
  The CMAKE_CXX_COMPILER:

    g++.exe

  is not a full path and was not found in the PATH.

  Tell CMake where to find the compiler by setting either the environment
  variable "CXX" or the CMake cache entry CMAKE_CXX_COMPILER to the full path
  to the compiler, or to the compiler name if it is in the PATH.


CMake Error at CMakeLists.txt:9 (project):
  The CMAKE_C_COMPILER:

    gcc.exe

  is not a full path and was not found in the PATH.

  Tell CMake where to find the compiler by setting either the environment
  variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
  the compiler, or to the compiler name if it is in the PATH.


-- Configuring incomplete, errors occurred!
See also "D:/dev/makhber/build/test/CMakeFiles/CMakeOutput.log".
See also "D:/dev/makhber/build/test/CMakeFiles/CMakeError.log".
```
but I think this should be reported to CMake instead because it succeeds with Ninja which means they assume "MinGW"==GCC.

**Comment by MehdiChinoune** [comment]
> Do notice too that this affects Boost builds with the `toolset=gcc` parameter; we now need to add another override to target Clang and its specific suffix to make our ExternalProject work.

you may take a look at https://github.com/msys2/MINGW-packages/pull/10719

**Comment by MehdiChinoune** [comment]
with "MinGW Makefiles" generator, cmake only searches for gcc/g++ https://gitlab.kitware.com/cmake/cmake/-/blob/master/Source/cmGlobalMinGWMakefileGenerator.cxx#L32
I think we could patch it to search for cc/c++ instead until they fix the issue.

**Comment by amyspark** [comment]
This issue also impacts building cargo-c from scratch, as the linker is missing. I tried both with the official (rustup) and the msys supplied version:

```
error: linker `x86_64-w64-mingw32-gcc` not found
  |
  = note: El sistema no puede encontrar el archivo especificado. (os error 2)

error: could not compile `winapi-x86_64-pc-windows-gnu` due to previous error
warning: build failed, waiting for other jobs to finish...
error: failed to compile `cargo-c v0.8.1+cargo-0.53 (E:\krita-win\b_deps_clang64\ext_heif\ext_cargo_c-prefix\src\ext_cargo_c)`, intermediate artifacts can be found at `E:\krita-win\b_deps_clang64\ext_heif\ext_cargo_c-prefix\src\ext_cargo_c\target`
```

**Comment by MehdiChinoune** [comment]
> This issue also impacts building cargo-c from scratch, as the linker is missing. I tried both with the official (rustup) and the msys supplied version:

Fixed in #10789

**Comment by jeremyd2019** [comment]
for comparison, llvm-mingw provides gcc/g++ aliases for clang
https://github.com/mstorsjo/llvm-mingw/blob/26dd3e46fe630495a174b58273c45ec72b306307/install-wrappers.sh#L88

**Comment by jeremyd2019** [comment]
perl appears to depend on the gcc/g++ names.  Even patching the makefile to not use gcc/g++ that succeeds in building miniperl, but once it gets to building extensions it passes a mix of unix and msvc style parameters to link, resulting in errors.

**Comment by jeremyd2019** [comment]
That was with using clang/clang++ names.  Switching to cc/c++ names caused it to fail earlier.  Perl's build system, at least for mingw, is a hot mess.

**Comment by jeremyd2019** [comment]
I tacked on in #10824 
> I also made a new split package on mingw-w64-clang that contains the gcc aliases (mingw-w64-clang-*-gcc-compat). This inherits the '-toolchain' group from the top-level of the PKGBUILD so hopefully users who want to use tools that assume gcc will be able to continue to use them, and packages whose logic is tied to the name 'gcc' like perl can makedepend on it explicitly.

Hopefully this will help.  Note the issue that PR was created to address will require two builds of clang, so it may be a while before that package is deployed.

**Comment by lazka** [comment]
[mingw-w64-clang-x86_64-gcc-compat](https://packages.msys2.org/package/mingw-w64-clang-x86_64-gcc-compat?repo=clang64) is now in the repo

**Comment by amyspark** [comment]
Fixed in #10824
